src/cunumeric/stat/histogram.cu

```diff
@@ -0,0 +1,235 @@
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size = src_rect.hi - src_rect.lo + 1;
+  return std::make_tuple(src_size, src_ptr);
+}
+// accessor copy utility:
+//
+template <typename VAL>
+std::tuple<size_t, Buffer<VAL>, const VAL*> make_accessor_copy(const AccessorRO<VAL, 1>& src_acc,
+                                                               const Rect<1>& src_rect)
+{
+  size_t src_strides[1];
+  const VAL* src_ptr = src_acc.ptr(src_rect, src_strides);
+  assert(src_strides[0] == 1);
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size      = src_rect.hi - src_rect.lo + 1;
+  Buffer<VAL> src_copy = create_buffer<VAL>(src_size, Legion::Memory::Kind::GPU_FB_MEM);
+  return std::make_tuple(src_size, src_copy, src_ptr);
+}
+}  // namespace detail
+
+template <Type::Code CODE>
+struct HistogramImplBody<VariantKind::GPU, CODE> {
+  using VAL = legate_type_of<CODE>;
+
+  // for now, it has been decided to hardcode these types:
+  //
+  using BinType    = double;
+  using WeightType = double;
+
+  // in the future we might relax relax that requirement,
+  // but complicate dispatching:
+  //
+  // template <typename BinType = VAL, typename WeightType = VAL>
+  void operator()(const AccessorRO<VAL, 1>& src,
+                  const Rect<1>& src_rect,
+                  const AccessorRO<BinType, 1>& bins,
+                  const Rect<1>& bins_rect,
+                  const AccessorRO<WeightType, 1>& weights,
+                  const Rect<1>& weights_rect,
+                  const AccessorRD<SumReduction<WeightType>, true, 1>& result,
+                  const Rect<1>& result_rect) const
+  {
+    auto stream          = get_cached_stream();
+    cudaStream_t stream_ = static_cast<cudaStream_t>(stream);
+
+    auto&& [src_size, src_copy, src_ptr] = detail::make_accessor_copy(src, src_rect);
+    CHECK_CUDA(cudaMemcpyAsync(
+      src_copy.ptr(0), src_ptr, src_size * sizeof(VAL), cudaMemcpyDeviceToDevice, stream));
+
+    auto&& [weights_size, weights_copy, weights_ptr] =
+      detail::make_accessor_copy(weights, weights_rect);
+    CHECK_CUDA(cudaMemcpyAsync(weights_copy.ptr(0),
+                               weights_ptr,
+                               weights_size * sizeof(WeightType),
+                               cudaMemcpyDeviceToDevice,
+                               stream));
+
+    auto&& [bins_size, bins_ptr] = detail::get_accessor_ptr(bins, bins_rect);
+
+    auto num_intervals = bins_size - 1;
+    Buffer<WeightType> local_result =
+      create_buffer<WeightType>(num_intervals, Legion::Memory::Kind::GPU_FB_MEM);
+
+    WeightType* local_result_ptr = local_result.ptr(0);
+
+    auto&& [global_result_size, global_result_ptr] = detail::get_accessor_ptr(result, result_rect);
+
+    CHECK_CUDA(cudaStreamSynchronize(stream));
```

Comment:
Let's replace these with `CHECK_CUDA_STREAM(stream)`, which will only sync the stream in debug mode.

Reply:
Why would we want the stream synchronized _only_ in debug mode? We essentially need a barrier, here, that says: okay let's wait until all async copies are finished before using the pointers behind them. 

That needs to happen regardless of debug/production mode. Now, since the copies are device-to-device on the given stream and the calls that follow are device calls on the same stream, we might get away with it (b/c no host-only operation is involved), but in general it's not good practice. Are we sure we want that?